wordpress-mobile/WordPress-Android
# Small Reader Improvements: Update Reader Tracking Properties

Fixes #14069

This tracking related updates were requested in order to enhance the data for the `Reader`.

After consulting the associated spreadsheet and investigating all `Reader` related tracking events it was decided to add all three main `Reader` tracking properties, wherever possible. Those are the:

- Blog Id (as `blog_id`, with a `numeric` value)
- Feed Id (as `feed_id`, with a `numeric` value) - This is based on @ashiagr feedback! 🙏 
- Is Followed (as `follow`, with `boolean` value)
- Source (as `source`, with a `string` value)

```
The `follow` property derived from:

- ReaderBlog.isFollowing, or
- ReaderPost.isFollowedByCurrentUser

I choose one over the other based on the context and it being available within that context.
I am not sure whether we should be using `ReaderBlog.isFollowing` over the 
`ReaderPost.isFollowedByCurrentUser` in all instances. Please code review and test this 
functionality in more depth to verify correctness.
```

```
The `source` property holds one of the below values:

- following
- discover
- liked
- saved
- custom
- a8c
- p2
- subscriptions
- search
- site_preview
- tag_preview
- post_detail
- comment
- user
- stats
- notification
- activity_log_detail
- unknown

Also, there is an extra source value of `post_list_saved_post_notice`, but this is an existing one.
```

As such please expect an analogous adjustment to what the spreadsheet dictated.

Below is a table with all the tracking events, their triggers and the tracking properties that were added (as `NEW`), already existed (as `EXISTS`) or not added (as `N/A`).

Number|Tracking Event | Trigger | Blog Id | Feed Id | Is Followed | Source
-------|-------|------|------|------|------|------
1|reader_discover_topic_tapped|When a user scrolls through the 'DISCOVER' tab, finds the 'You might like' card and clicks on a suggested 'topic' within that card.|~~N/A (*)~~|~~N/A (*)~~|~~N/A (*)~~|~~N/A (**)~~
2|reader_tag_previewed|When a user clicks on a 'topic' within a card.|~~N/A (*)~~|~~N/A (*)~~|~~N/A (*)~~|**NEW**
3|reader_post_card_tapped|When a user clicks on a card.|**NEW**|**NEW**|**NEW**|**NEW**
4|reader_article_opened|When a user clicks on a card and the article is shown.|*EXISTS*|*EXISTS*|**NEW**|~~N/A (***)~~
5|reader_article_rendered|When a user clicks on a card and the article is rendered.|*EXISTS*|*EXISTS*|**NEW**|~~N/A (***)~~
6|reader_blog_previewed|When a user clicks on a card's site.|**NEW**|**NEW**|**NEW**|**NEW**
7|reader_site_followed|When a user clicks on a card and then clicks on the site 'Follow' button within the card.|*EXISTS*|**NEW**|~~N/A~~|**NEW**
8|reader_site_unfollowed|When a user clicks on a card and then clicks on the site 'Unfollow' button within the card.|*EXISTS*|**NEW**|~~N/A~~|**NEW**
9|reader_suggested_site_toggle_follow|When a user is within the 'DISCOVER' tab, finds a suggested 'site' and then clicks the 'Follow/Unfollow' button.|*EXISTS*|**NEW**|*EXISTS*|~~N/A (**)~~
10|reader_reader_tag_followed|When a user is within the 'DISCOVER' tab, finds a suggested 'topic' and then clicks the 'Follow' button.|~~N/A (*)~~|~~N/A (*)~~|~~N/A~~|**NEW**
11|reader_reader_tag_unfollowed|When a user is within the 'DISCOVER' tab, finds a suggested 'topic' and then clicks the 'Unfollow' button.|~~N/A (*)~~|~~N/A (*)~~|~~N/A~~|**NEW**
12|reader_site_shared|When a user clicks on a card's site and then clicks on the share button within the site.|**NEW**|**NEW**|**NEW**|**NEW**
13|item_shared_reader|When a user clicks on card's menu (see ellipsis), and then clicks on the share menu option.|*EXISTS*|**NEW**|**NEW**|**NEW**
14|reader_post_saved|When a user clicks on card's save button.|**NEW**|**NEW**|**NEW**|**NEW**
15|reader_post_unsaved|When a user clicks on card's unsave button.|**NEW**|**NEW**|**NEW**|**NEW**
16|reader_saved_post_opened|When a user opens a saved card from any tab.|**NEW**|**NEW**|**NEW**|**NEW**
17|reader_article_liked|When a user clicks on card's like button.|*EXISTS*|*EXISTS*|**NEW**|**NEW**
18|reader_article_unliked|When a user clicks on card's unlike button.|*EXISTS*|*EXISTS*|**NEW**|**NEW**
19|reader_article_detail_liked|When a user clicks on a card and then clicks on the like button within the card.|*EXISTS*|*EXISTS*|**NEW**|**NEW**
20|reader_article_detail_unliked|When a user clicks on a card and then clicks on the unlike button within the card.|*EXISTS*|*EXISTS*|**NEW**|**NEW**

```
* - This means that this is a `Topic` or `Tag` and as such the `Blog Id` or `Feed Id` property is not applicable to it.
** - This means that this tracking event only occurs on the `DISCOVER` tab and as such the `Source` property is omitted.
*** - This means that this tracking event only occurs on the `POST DETAIL` tab (ish) and as such the `Source` property is omitted.
```

Also, as part of this `Reader` tracking properties update, the below enhancements occurred:
- For every `Reader` related screen and classes, the `ReaderTracker` wrapper class replaced all other tracking related classes (like `AnalyticsTrackerWrapper`, `AnalyticsUtilsWrapper`, `AnalyticsUtils`, etc). This is done in order to consolidate all `Reader` related tracking functionality in one place and make it reusable on all places.
- For every `Reader` related screen and classes that the `ReaderTracker` wrapper was utilised, a automatic refactoring occurred as an initial step to make the code easier to work with (like removing all possible warning, optimizing `Reader` related imports, etc).
- A lot of `Reader` tracking related functionality was optimized (unnecessary functions were deleted, nullability annotations were added, unused code was removed, etc).

Warning (⚠️):
- To review this PR please do a commit-by-commit review from within the IDE.
- Make sure to spend some time on the `Feature` related commit and skip the `Refactor` related commits.

To test:
- Launch app and go to `My Site` tab.
- Go to: `My Site` -> `Toolbar Avatar` -> `App Settings` -> `Privacy settings`
- Switch on `Collect information` to enable tracking on your device.
- Navigate to the `Reader` tab.
- Follow the table above, per tracking event, trigger the event and verify that the newly added properties are tracked. 

PR submission checklist:
